src/kulprit/projection/solver.py

```diff
@@ -1,120 +1,147 @@
-        return theta_perp, final_loss
-
-    def solve_dispersion(
-        self, theta_perp: torch.tensor, X_perp: torch.tensor
-    ) -> torch.tensor:
-        """Analytic projection of the model dispersion parameters.
-
-        Args:
-            theta_perp (torch.tensor): A PyTorch tensor of the restricted
-                parameter draws
-            X_perp (np.ndarray): The design matrix of the restricted model we
-                are projecting onto
-
-        Returns:
-            torch.tensor: The restricted projections of the dispersion parameters
+        input_array = input_array[np.isfinite(input_array)].astype("float64")
+        if len(input_array) == 0:
+            return np.nan
+        else:
+            return np.mean(input_array)
+
+    def solve(
+        self,
+        term_names: list,
+        num_steps: Optional[int] = 5_000,
+        obj_n_mc: Optional[float] = 10,
+    ) -> SubModel:
+        """The primary projection method in the procedure.
+
+        The projection is performed with a mean-field approximation to variational
+        inference rather than concatenating posterior draw-wise optimisation
+        solutions as is suggested by Piironen (2018).
         """
 
-        # compute the solution and return
-        solution = self.family.solve_dispersion(theta_perp=theta_perp, X_perp=X_perp)
-        return solution
+        # if projecting onto the reference model, simply return it
+        if set(term_names) == set(self.ref_model.common_terms.keys()):
+            return SubModel(
+                model=self.ref_model,
+                idata=self.ref_idata,
+                elbo=np.inf,
+                size=len(self.ref_model.common_terms),
+                term_names=term_names,
+            )
+
+        # build restricted model
+        new_model = self._build_restricted_model(term_names=term_names)
+        new_pymc_model = new_model.backend.model
+
+        # perform mean-field MLE
+        with new_pymc_model:
+            approx = pm.MeanField()
+            inference = pm.KLqp(approx, beta=0.0)
+            mean_field = inference.fit(n=num_steps, obj_n_mc=obj_n_mc, progressbar=False)
+
+        # compute the LOO-CV predictive performance of the submodel
+        num_draws = (
+            self.ref_idata.posterior.dims["chain"]
+            * self.ref_idata.posterior.dims["draw"]
+        )
+        trace = mean_field.sample(num_draws, return_inferencedata=False)
+
+        # first obtain the aesara observed RVs
+        new_obs_rvs = new_pymc_model.named_vars[self.response_name]
+        old_obs_rvs = self.pymc_model.named_vars[self.response_name]
+        # and then we replace the observations in the new model for those in the reference model
+        new_pymc_model.rvs_to_values[new_obs_rvs] = self.pymc_model.rvs_to_values[
+            old_obs_rvs
+        ]
+        new_idata = pm.to_inference_data(
+            trace=trace, model=new_pymc_model, log_likelihood=True
+        )
+
```

Comment:
tidy up

```suggestion
        coords_to_drop = [dim for dim in new_idata.posterior.dims if dim.endswith("_dim_0")]
        new_idata.posterior = new_idata.posterior.squeeze(coords_to_drop).reset_coords(
            coords_to_drop, drop=True
        )
```
taken from taken from https://github.com/bambinos/bambi/blob/2287de43308338234a1cdd0edf2078205bc077a2/bambi/backend/pymc.py#L355